protocols/dcutr/src/handler/relayed.rs

```diff
@@ -0,0 +1,380 @@
+// Copyright 2021 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! [`ProtocolsHandler`] handling relayed connection potentially upgraded to a direct connection.
+
+use crate::protocol;
+use futures::future::{BoxFuture, FutureExt};
+use futures::stream::{FuturesUnordered, StreamExt};
+use instant::Instant;
+use libp2p_core::either::{EitherError, EitherOutput};
+use libp2p_core::multiaddr::Multiaddr;
+use libp2p_core::upgrade::{self, DeniedUpgrade, NegotiationError, UpgradeError};
+use libp2p_core::ConnectedPoint;
+use libp2p_swarm::protocols_handler::{InboundUpgradeSend, OutboundUpgradeSend};
+use libp2p_swarm::{
+    KeepAlive, NegotiatedSubstream, ProtocolsHandler, ProtocolsHandlerEvent,
+    ProtocolsHandlerUpgrErr, SubstreamProtocol,
+};
+use std::collections::VecDeque;
+use std::fmt;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+pub enum Command {
+    Connect {
+        obs_addrs: Vec<Multiaddr>,
+        attempt: u8,
+    },
+    AcceptInboundConnect {
+        obs_addrs: Vec<Multiaddr>,
+        inbound_connect: protocol::inbound::PendingConnect,
+    },
+    /// Upgrading the relayed connection to a direct connection either failed for good or succeeded.
+    /// There is no need to keep the relayed connection alive for the sake of upgrading to a direct
+    /// connection.
+    UpgradeFinishedDontKeepAlive,
```

Comment:
> Do we have a mechanism that will restart dcutr?

DCUtR has an internal mechanism that retries twice, resulting in 3 upgrade attempts. If all 3 fail it does not retry again.

> But it makes me wonder what happens if the relayed connection stays up, but the direct connection fails?

DCUtR itself will not keep the relayed connection alive. Though other protocols might still use it up until the time or bandwidth limit of the relayed connection is hit and thus the connection breaks.

I have been thinking about whether upper-level protocols need this to be more explicit, in other words whether they need to know that they are sending data over a relayed connection which might break at any point in time. Upper-level protocols will need to be able to cope with connection failures anyways. In the current state of DCUtR we expect upper-level protocols to handle upgrade failures and relay connection terminations with that same mechanism.

I am happy for alternative suggestions. Would be cool to be able to provide more information to upper-level applications. Though unless done in a generic way, I am not sure it is worth the additional complexity.

@MarcoPolo what do you think?

Reply:
> DCUtR itself will not keep the relayed connection alive. Though other protocols might still use it up until the time or bandwidth limit of the relayed connection is hit and thus the connection breaks.

I think this makes sense. Here's a use case I'm thinking of as a user of this behavior:

1. I use this behavior as part of my root behavior.
2. I successfully upgrade my connection to a direct one.
3. Some time passes.
4. The direct connection fails (this is fine and expected).
5. How do I try to get another direct connection?

Does the other peer have to close all existing relayed connections with me and start another one so my `inject_connection_established` callback can happen?
Or maybe (in the future, not this PR) there should be method in the behavior that allows one to try to re-setup a dcutr dance.

Reply:
> How do I try to get another direct connection?

For now you would need to initiate a new relayed connection, which would lead to the destination of the relayed connection to initiate a _direct connection upgrade_.

>  Or maybe (in the future, not this PR) there should be method in the behavior that allows one to try to re-setup a dcutr dance.

Over an existing relayed connection, potentially the relayed connection previously used? We would first have to update the [DCUtR specification](https://github.com/libp2p/specs/blob/master/relay/DCUtR.md) as it only details a _direct connection upgrade_ when a new relayed connection is established.

We could e.g. agree that the listening side of the relayed connection is allowed to initiate a new _direct connection upgrade_ in case the first direct connection dies.

---

From a more general perspective, something that might be worth noting here:

Say a `ProtocolsHandler` returned `KeepAlive::No` on `ProtocolsHandler::connection_keep_alive`. It basically signals: "I don't need the connection right now. I am fine with others using it.".

Returning `KeepAlive::No` does not stop the `ProtocolsHandler` from continuing to operate and it will continue to be `poll`ed by the task of the connection. E.g. `libp2p-ping` always returns `KeepAlive::No` as it is just an auxiliary protocol with no utility in itself.

In case the connection is kept alive by some other protocol, in other words in case the connection has not been closed, the `ProtocolsHandler` is free to change its mind, i.e. return `KeepAlive::Yes` again.

Reply:
> For now you would need to initiate a new relayed connection, which would lead to the destination of the relayed connection to initiate a direct connection upgrade.

ah this makes sense. I think this is okay to do as a user.

---

> Returning KeepAlive::No does not stop the ProtocolsHandler from continuing to operate and it will continue to be polled by the task of the connection. E.g. libp2p-ping always returns KeepAlive::No as it is just an auxiliary protocol with no utility in itself.

Ah! thanks! that makes sense. The `KeepAlive` is a signal for the connection to close, and the handler itself is **only closed** when the [connection is closed](https://github.com/libp2p/rust-libp2p/blob/master/core/src/connection/pool/task.rs#L204). A connection will only be closed (barring failure modes) when all handlers using that connection return `KeepAlive::No` in the case of a `ProtocolsHandlerSelect` [code](https://github.com/libp2p/rust-libp2p/blob/master/swarm/src/protocols_handler/select.rs#L319).

If this were the only handler associate with the connection (only one behavior), returning `KeepAlive::No` would close the connection and drop the handler. But since usually there are many behaviours in a hierarchy, the top level handler in the hierarchy returns the `cmp::max` of all handlers in the hierarchy.

This is pretty nice since this means the swarm code doesn't know the difference between the handlers. In its perspective there's only a single handler for the connection.

Reply:
Marking as resolved. Please re-open in case I am missing something.